### PR TITLE
Support new kernels

### PR DESCRIPTION
**Pull request by yanivagman**

### CODE REVIEW

**Comment by grantseltzer** [comment]
Can confirm this works on 5.3 :+1:

**Comment by yanivagman** [comment]
Thanks @grantseltzer , actually I just tested it with kernel 4.19 and it seems to work for pid, tid, etc...
However, for some reason it seems that the strings are omitted (getting a number instead). For example in open syscall, I can't see the arguments. Do you have this issue too in 5.3?
I think that it's related to one of the fields in the context_t, which size may have changed... Need to do some more investigation on this

**Comment by grantseltzer** [comment]
I see some:

```
31744.766047   host             4026531840   4026531836   1000   execve           zsh              41688  41688  41685  0               /usr/libexec/grepconf.sh  
31744.760897   host             4026531840   4026531836   1000   openat           zsh              41685  41685  41684  3            1149124440  O_RDONLY
```

In the above `openat` call, I suspect the number is an address that's trying to be interpreted as a string in decimal. 

I haven't ever run it on a 4.18 box so haven't seen it otherwise.

**Comment by yanivagman** [comment]
Thanks @grantseltzer , you should actually get the full strings given to open/openat - this is a bug and not an address.
I'll try to figure this out now that I updated my kernel to 4.19

**Comment by grantseltzer** [comment]
@yanivagman I've been trying to figure out that bug the past couple of days. [This](https://stackoverflow.com/questions/54963344/bpf-kprobe-socket-related-syscall-read-socket-detail-fail) question on SO describes the same problem. 

Perhaps we can make use of a higher level kernel function instead of the system call directly. I've been trying to trace the `seccomp` syscall but was having this issue, instead of i'm going to try tracing `do_seccomp()`.

**Comment by yanivagman** [comment]
@grantseltzer I also tried to figure this out, and found that syscall probes should be defined with syscall__ prefix (https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#8-system-call-tracepoints) . The reason for this seems to be some ABI change (https://lore.kernel.org/patchwork/patch/937552/) which is set by defining a config in the kernel CONFIG_ARCH_HAS_SYSCALL_WRAPPER, starting from kernel 4.17 (I guess, as patch is from March 2018). Looking at BCC code, I see that they are taking this into account when using the syscall__ prefix (https://github.com/iovisor/bcc/blob/149c1c8857652997622fc2a30747a60e0c9c17dc/src/cc/frontends/clang/b_frontend_action.cc#L678).

I tried to change the syscall probes to use syscall__ prefix (on branch test_syscall_on_new_kernels), and now I can see the strings in execve syscall, but it doesn't seem to fix other syscalls (e.g. open, openat, lstat...) - I suspect this might be a bug in BCC.

**Comment by yanivagman** [comment]
Indeed this seems to be BCC issue.
The issue is caused because of using the macro PT_REGS_PARM2 (and maybe other macros too) is not functioning as expected in new kernels for syscalls.
I opened a bug in bcc for that:
https://github.com/iovisor/bcc/issues/2628

**Comment by yanivagman** [comment]
Added workaround for bcc macros bug. It should now work as expected on new kernels

File: tracee/container_event_monitor_ebpf.c [hunk 1]
```diff
@@ -476,6 +483,26 @@ static __always_inline u32 get_task_ppid(struct task_struct *task)
...
+    args->args[4] = PT_REGS_PARM5(ctx);
+    args->args[5] = PT_REGS_PARM6(ctx);
+#else
+    struct pt_regs * ctx2 = (struct pt_regs *)ctx->di;
+    bpf_probe_read(&args->args[0], sizeof(args->args[0]), &ctx2->di);
+    bpf_probe_read(&args->args[1], sizeof(args->args[1]), &ctx2->si);
+    bpf_probe_read(&args->args[2], sizeof(args->args[2]), &ctx2->dx);
+    bpf_probe_read(&args->args[3], sizeof(args->args[3]), &ctx2->r10);
+    bpf_probe_read(&args->args[4], sizeof(args->args[4]), &ctx2->r8);
+    bpf_probe_read(&args->args[5], sizeof(args->args[5]), &ctx2->r9);
```
Comment (grantseltzer) [inline tracee/container_event_monitor_ebpf.c]: Ah, so you're just manually reading from the registers? Does that macro provide anything besides that? If not, why bother checking kernel version?

File: tracee/container_event_monitor_ebpf.c [hunk 2]
```diff
@@ -476,6 +483,26 @@ static __always_inline u32 get_task_ppid(struct task_struct *task)
...
+{
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 17, 0)
+    args->args[0] = PT_REGS_PARM1(ctx);
+    args->args[1] = PT_REGS_PARM2(ctx);
+    args->args[2] = PT_REGS_PARM3(ctx);
+    args->args[3] = PT_REGS_PARM4(ctx);
+    args->args[4] = PT_REGS_PARM5(ctx);
+    args->args[5] = PT_REGS_PARM6(ctx);
+#else
+    struct pt_regs * ctx2 = (struct pt_regs *)ctx->di;
```
Comment (grantseltzer) [inline tracee/container_event_monitor_ebpf.c]: Why is this needed?

File: tracee/container_event_monitor_ebpf.c (hunk 1 above)
Comment (yanivagman) [reply tracee/container_event_monitor_ebpf.c]: The macro does only this. The reason for the kernel version check is exactly because of the problem we saw between kernel versions. In kernels before 4.17, ctx has a different structure for systemcalls in x86 architecture - it was saved directly on ctx as pt_regs struct. From kernel 4.17, a new level of indirection was added, so now we have to first derefernce ctx->di, and only then we have pt_regs struct which can further dereference for getting the args

File: tracee/container_event_monitor_ebpf.c (hunk 2 above)
Comment (yanivagman) [reply tracee/container_event_monitor_ebpf.c]: Explained in my comment above

File: tracee/container_event_monitor_ebpf.c (hunk 1 above)
Comment (grantseltzer) [reply tracee/container_event_monitor_ebpf.c]: Interesting! Thank you for the info!

**Comment by simar7** [comment]
hey @yanivagman I tried to run it on a Ubuntu distro running the following:

```
vagrant@vagrant:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.2 LTS
Release:	18.04
Codename:	bionic
```

with the following kernel:

```
vagrant@vagrant:~$ uname -a
Linux vagrant 5.0.0-20-generic #21~18.04.1-Ubuntu SMP Thu Jun 27 04:04:37 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```


I see the following:
```
vagrant@vagrant:~/repos/contracer$ sudo ./start.py -c -e execve
TIME(s)        UTS_NAME         MNT_NS       PID_NS       UID    EVENT            COMM             PID    TID    PPID   RET          ARGS
446.167336     8a50ecefc348     4026532197   4026532200   0      execve           runc:[2:INIT]    1      1      3641   0            /hello
454.561546     7218f59d6baf     4026532197   4026532200   0      execve           runc:[2:INIT]    1      1      3795   0            /bin/bash
454.564486     7218f59d6baf     4026532197   4026532200   0      execve           bash             7      7      6      0            /usr/bin/groups
454.572340     7218f59d6baf     4026532197   4026532200   0      execve           bash             9      9      8      0            /usr/bin/dircolors -b
576.112377     7218f59d6baf     4026532197   4026532200   0      execve           bash             10     10     1      0            /bin/ls --color=auto -badoptions
```

I don't see ```UID``` and ```RET``` being populated. You can see the last `ls -badoptions` command I ran was with bad params returning an error code of `2`.

Is this expected?

**Comment by yanivagman** [comment]
> I don't see `UID` and `RET` being populated. You can see the last `ls -badoptions` command I ran was with bad params returning an error code of `2`.
> 
> Is this expected?

execve syscall is a special case as it destroys the address space of the calling process and initialize it with a new binary image. In order to be able to read all the arguments before the address space is destroyed, we send the message to userspace in syscall entry, and not exit. For this reason, if execve fails, you should get another line in the output indicating that, and not in the same line.
That being said, I don't see a problem with your output. I can see a column for UID with value 0 
(root user), and execve shouldn't fail with bad parameters - the program will start and exit with an error code. So you probably referring to the error code value, which we don't support here. We do support the return value of execve, which is indeed 0.

**Comment by simar7** [comment]
> > I don't see `UID` and `RET` being populated. You can see the last `ls -badoptions` command I ran was with bad params returning an error code of `2`.
> > Is this expected?
> 
> execve syscall is a special case as it destroys the address space of the calling process and initialize it with a new binary image. In order to be able to read all the arguments before the address space is destroyed, we send the message to userspace in syscall entry, and not exit. For this reason, if execve fails, you should get another line in the output indicating that, and not in the same line.
> That being said, I don't see a problem with your output. I can see a column for UID with value 0
> (root user), and execve shouldn't fail with bad parameters - the program will start and exit with an error code. So you probably referring to the error code value, which we don't support here. We do support the return value of execve, which is indeed 0.

Got it. Thanks.
